### PR TITLE
Revive aliases feature

### LINKED ISSUE #1722: Unable to specify alias
### Actual behavior
Hey,
I'm using the latest version 0.14 and somehow the alias feature has vanished. I'm unable to define or select an alias for my account.

Best regards,
Jonas

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/73573289-unable-to-specify-alias?utm_campaign=plugin&utm_content=tracker%2F44154351&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F44154351&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by IchbinkeinReh**
Fixes #1722

### CODE REVIEW

File: l10n/de.js [hunk 1]
```diff
@@ -11,6 +11,7 @@ OC.L10N.register(
     "Auto" : "Automatisch",
     "Name" : "Name",
     "Mail Address" : "E-Mail-Adresse",
+    "Aliases" : "Aliase",
```
Comment (ChristophWurst) [inline l10n/de.js:14]: Please revert any l10n updates. These are done via Transifex and a cron job, not manually.

File: lib/Db/Alias.php [hunk 2]
```diff
@@ -49,4 +51,13 @@ public function __construct() {
...
 		$this->addType('alias', 'string');
 	}
+
+	public function jsonSerialize() {
+		$result = [
+			'id' => $this->getId(),
+			'name' => $this->getName(),
+			'alias' => $this->getAlias()
+		];
+		return $result;
```
Comment (ChristophWurst) [inline lib/Db/Alias.php:61]: no need for the variable, return it directly :)

File: lib/Db/Alias.php [hunk 3]
```diff
@@ -49,4 +51,13 @@ public function __construct() {
 		$this->addType('name', 'string');
 		$this->addType('alias', 'string');
 	}
+
+	public function jsonSerialize() {
```
Comment (ChristophWurst) [inline lib/Db/Alias.php:55]: you mad add a `: array` return type hint here

File: src/components/AliasSettings.vue [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+  -
+  - You should have received a copy of the GNU Affero General Public License
+  - along with this program.  If not, see <http://www.gnu.org/licenses/>.
+  -->
+
+<template>
+	<div id="aliases" class="section">
+		<h2>{{ t('mail', 'Aliases') }}</h2>
+		<div id="aliases-list">
+			<table class="grid">
```
Comment (ChristophWurst) [inline src/components/AliasSettings.vue:24]: @jancborchardt is there a way around table?

File: src/components/AliasSettings.vue [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	props: {
+		account: {
+			type: Object,
+			required: true,
+		},
+	},
+	data() {
+		return {
+			loading: false,
+			aliases: this.account.aliases,
```
Comment (ChristophWurst) [inline src/components/AliasSettings.vue:79]: this should be a `computed` property so it updates with vuex mutations

File: src/components/AliasSettings.vue [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+			alias: {aliasName: '', alias: ''},
+		}
+	},
+	methods: {
+		deleteAlias(alias) {
+			this.loading = true
+
+			this.$store
```
Comment (ChristophWurst) [inline src/components/AliasSettings.vue:87]: I would recommend `async`/`await` if you are familiar with that, you can use promises as well if you are more comfortable with that

File: src/components/Composer.vue [hunk 7]
```diff
@@ -335,7 +335,23 @@ export default {
 	},
 	computed: {
 		aliases() {
-			return this.$store.getters.accounts.filter((a) => !a.isUnified)
+			let cnt = 0
+			const aliases = []
+			const accounts = this.$store.getters.accounts.filter((a) => !a.isUnified)
+			accounts.forEach((account) => {
+				account.aliasId = null
```
Comment (ChristophWurst) [inline src/components/Composer.vue:342]: :warning: this is :fish:y. You should not mutate any data from the store. but you assign new props here. Vue will definitely warn you about this in the dev build. Any change of the state should go through an action and a mutation.

perhaps what you want to do is `.map(...)` the accounts to a new structure, as in, create new object

File: src/views/AccountSettings.vue [hunk 8]
```diff
@@ -51,6 +53,7 @@ export default {
 		EditorSettings,
 		Navigation,
 		SignatureSettings,
+		AliasSettings,
```
Comment (ChristophWurst) [inline src/views/AccountSettings.vue:56]: you'll get bonus points if you arrange the props alphabetically :wink:

**Comment by ChristophWurst** [comment]
Btw @IchbinkeinReh I've added you as collaborator on the repo :)

**Comment by jancborchardt** [comment]
Nice work! :) Some design feedback:

- [ ] We don’t need a separate section here, we can simply integrate it into the name + email display at the top:
    image
- [ ] This also means it doesn’t need to be a table but can simply be a list → also referring to what @ChristophWurst said above.
- [ ] The name input should go first, and it can be prefilled with the display name, with the email field then being focused.

File: l10n/de.js (hunk 1 above)
Comment (IchbinkeinReh) [reply l10n/de.js:14]: Done

File: lib/Db/Alias.php (hunk 2 above)
Comment (IchbinkeinReh) [reply lib/Db/Alias.php:61]: Done

File: lib/Db/Alias.php (hunk 3 above)
Comment (IchbinkeinReh) [reply lib/Db/Alias.php:55]: Done

File: src/views/AccountSettings.vue (hunk 8 above)
Comment (IchbinkeinReh) [reply src/views/AccountSettings.vue:56]: Done

**Comment by IchbinkeinReh** [comment]
@jancborchardt  What do you think about this?
ichbinkeinreh de_cloud_index php_apps_mail_ (1)
ichbinkeinreh de_cloud_index php_apps_mail_

File: src/components/AliasSettings.vue (hunk 6 above)
Comment (IchbinkeinReh) [reply src/components/AliasSettings.vue:87]: Thanks for the suggestion. Async await really makes the code much more readable

File: src/components/Composer.vue [hunk 9]
```diff
@@ -335,7 +335,24 @@ export default {
 	},
 	computed: {
 		aliases() {
-			return this.$store.getters.accounts.filter((a) => !a.isUnified)
+			let cnt = 0
+			const aliases = []
+			const accounts = this.$store.getters.accounts.filter((a) => !a.isUnified)
+			accounts.forEach((account) => {
```
Comment (ChristophWurst) [inline src/components/Composer.vue:341]: I have the feeling something like `flatMap` would be more appropriate here. Which attributes of the account/alias do we need? Is a full deep copy necessary or can we just build new objects from the existing data?

Comment (IchbinkeinReh) [reply src/components/Composer.vue:341]: The selectedAlias which contains the selected value from the to field is used in 10 places in the composer. But you are right, not all data from the account is needed. Only the following attributes are needed: editorMode, signature, id, aliasId

Comment (IchbinkeinReh) [reply src/components/Composer.vue:341]: @ChristophWurst BTW flatmap has no IE support is that a problem?

Comment (ChristophWurst) [reply src/components/Composer.vue]: You can use https://lodash.com/docs/4.17.15#flatMap but we also have Babel so it should polyfill if flatMap is an ES standard

**Comment by IchbinkeinReh** [comment]
@ChristophWurst Can this PR be merged now? Or didnt you have time to look into the changes yet?

**Comment by ChristophWurst** [comment]
I'll have a look very soon :pray:

**Comment by jancborchardt** [comment]
By the way @IchbinkeinReh we have a Nextcloud Talk instance with a channel for the Nextcloud Mail team – I can invite you there with a guest account if you like (via the email address + name from your commit)? :)

**Comment by IchbinkeinReh** [comment]
@jancborchardt Hey, thanks but its not necessary. I probably won't work at anything related to nextcloud mail the next time. Only wanted to revive the aliases and test my VueJS skills ;)
But if you guys need a helping hand I could do some paid contract work. Just contact me by mail if that's the case

**Comment by ChristophWurst** [comment]
@MarZi235 use https://help.nextcloud.com/c/apps/mail/35 for questions please.

**Comment by MarZi235** [comment]
Thank you, Christoph. Posted https://help.nextcloud.com/t/is-the-alias-feature-dropped/159972

**Pull request merged**